File: USACO/Training Website/Chapter 1/Section 1.2 - Ad Hoc/gift1.py
"""
ID: sunnyky1
LANG: PYTHON3
TASK: gift1
"""

with open("gift1.in", 'r') as fin:
    num_people = int(fin.readline().strip())
    people = {}
    for i in range(num_people):
        name = fin.readline().strip()
        people[name] = 0
    for i in range(num_people):
        giver = fin.readline().strip()
        money, split_count = [int(n) for n in fin.readline().strip().split()]
        if split_count != 0:
            money_per_person = money // split_count
            left_over = round(money % split_count)
            people[giver] = people[giver] - money + left_over
            for j in range(split_count):
                receiver = fin.readline().strip()
                people[receiver] += money_per_person

with open("gift1.out", 'w') as fout:
    for p in people:
        fout.write("{} {}\n".format(p, people[p]))

File: USACO/Training Website/Chapter 1/Section 1.2 - Ad Hoc/gift1.java
/*
ID: sunnyky1
LANG: JAVA
TASK: gift1
*/

import java.util.*;
import java.io.*;

public class gift1 {
	static HashMap<String, Integer> money;
	static String[] people;
	static int np;

	public static void main(String[] args) throws IOException {
		BufferedReader fin = new BufferedReader(new FileReader("gift1.in"));
		np = Integer.parseInt(fin.readLine());
		money = new HashMap<String, Integer>(np, 1);
		people = new String[np];
		for (int i = 0; i < np; i++) {
			people[i] = fin.readLine();
			money.put(people[i], 0);
		}
		// Simulate all gifts
		String giver, receiver;
		int amountToGive, numPeople;
		StringTokenizer st;
		for (int i = 0; i < np; i++) {
			giver = fin.readLine();
			st = new StringTokenizer(fin.readLine());
			amountToGive = Integer.parseInt(st.nextToken());
			numPeople = Integer.parseInt(st.nextToken());
			if (numPeople != 0) {
				amountToGive -= amountToGive % numPeople;
				money.put(giver, money.get(giver) - amountToGive);
				for (int j = 0; j < numPeople; j++) {
					receiver = fin.readLine();
					money.put(receiver, money.get(receiver) + amountToGive / numPeople);
				}
			}
		}
		fin.close();
		
		PrintWriter fout = new PrintWriter(new BufferedWriter(new FileWriter("gift1.out")));
		for (String person : people) {
			fout.println(person + " " + money.get(person));
		}
		fout.close();
	}

}
